tripleawarclub.org/public_html/dice/dice.class.php

```diff
@@ -0,0 +1,287 @@
+		$num_emails = $result->num_rows;
+
+		return ($num_emails == 1);
+	}
+
+	/**
+	 * runs any SQL query on the database
+	 * @param string $sql
+	 * @return mixed mysqlressource
+	 */
+	function runQuery($sql) {
+		$this->connectDatabase();
+
+		$result = $this->dbconn->query($sql) or exit("fatal error: " . $this->dbconn->error . "!");
+		return $result;
+	}
+
+	function runStatement($sth){
+		$this->connectDatabase();
+		$result = $sth->execute();
+		return $result;
+	}
+    function disconnectDatabase() {
+    	mysqli_close($this->dbconn);
+    }
+
+	////////////////////////////////
+	//        Encryption          //
+	////////////////////////////////
+	/**
+	 * returns the date and key
+	 * if no date is specified the latest key in key.dat will be returned
+	 */
+	function getEncryptionKey($date = null) {
+		// get old key
+		if($date) {
+			$dir = dirname(__FILE__);
+			$keyfile = fopen("$dir/keys/$date.dat", "r");
+		} else {
+			// get current key
+			if (!file_exists("key.dat")) {
+				$this->setEncryptionKey();
+			}
+			$keyfile = fopen("key.dat", "r");
+		}
+
+		if ($keyfile) {
+			$data = fread($keyfile, 8192);
+
+			$this->enc = unserialize($data);
+			fclose($keyfile);
+
+			return $this->enc;
+		} else {
+			exit("fatal error: Wrong date!");
+		}
+	}
+
+	function setEncryptionKey() {
+		$shouldWriteKey = true;
+		if (file_exists("key.dat")) {
+			$old = $this->getEncryptionKey();
+			$shouldWriteKey = copy("key.dat", "./keys/" . $old['date'] . ".dat");
+		}
+
+		if ($shouldWriteKey) {
+			$file = fopen("key.dat", "w");
+			$enc['key'] = $this->keygen();
+			$enc['date'] = $this->getDate();
+			$output = serialize($enc);
+			fputs($file, $output);
+			fclose($file);
+			chmod("key.dat", 0600);
+		}
+	}
+
+    function keygen() {
+		$tempstring = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+		for($length = 1; $length < 24; $length++) {
+			$temp = str_shuffle($tempstring);
+			$char = mt_rand(0, strlen($temp));
+			$pass .= $temp[$char];
+		}
+		return $pass;
+	}
+
+	function encrypt_data($input) {
+		$this->getEncryptionKey();
+		$string = serialize($input);
+
+		$td = mcrypt_module_open('rijndael-256', '', 'cfb', '');
+    	$out['iv'] = mcrypt_create_iv (mcrypt_enc_get_iv_size($td), MCRYPT_RAND);
+    	mcrypt_generic_init($td, $this->enc['key'], $out['iv']);
+    	$encrypted_data = mcrypt_generic($td, $string);
+    	mcrypt_generic_deinit($td);
+    	mcrypt_module_close($td);
+
+		$out['data'] = rawurlencode(base64_encode($encrypted_data));
+		$out['iv'] = rawurlencode(base64_encode($out['iv']));
+
+		return $out;
+	}
+
+	function decrypt_data($date, $iv, $encrypted_data) {
+		$iv = base64_decode(rawurldecode($iv));
+		$encrypted_data = base64_decode(rawurldecode($encrypted_data));
+
+		if ($date != $this->getDate()) {
+			$encrypt_key = $this->getEncryptionKey($date);
+		} else {
+			$encrypt_key = $this->getEncryptionKey();
+		}
```

Comment:
This if else statement can be shortened into a short assignment I forgot what it's actually called: `$encrypt_key = $this->getEncryptionKey(($date !== $this->getDate()) ? $date : null);`